statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,369 @@
+	for tblID, element := range s.elements {
+		newCache.elements[tblID] = element
+	}
+	return newCache
+}
+
+func (s *statsInnerCache) onEvict(tblID int64) {
+	element, exist := s.elements[tblID]
+	if !exist {
+		return
+	}
+	element.tblMemUsage = element.tbl.MemoryUsage()
+}
+
+func (s *statsInnerCache) freshTableCost(tblID int64, element *lruMapElement) {
+	element.tblMemUsage = element.tbl.MemoryUsage()
+	for idxID, idx := range element.tbl.Indices {
+		s.lru.put(tblID, idxID, idx, element.tblMemUsage.IndicesMemUsage[idxID], true, false)
+	}
+}
+
+func (s *statsInnerCache) capacity() int64 {
+	return s.lru.capacity
+}
+
+func (c *innerItemLruCache) get(tblID, id int64) (*lruCacheItem, bool) {
+	v, ok := c.elements[tblID]
+	if !ok {
+		return nil, false
+	}
+	ele, ok := v[id]
+	if !ok {
+		return nil, false
+	}
+	c.cache.MoveToFront(ele)
+	return ele.Value.(*lruCacheItem), true
+}
+
+func (c *innerItemLruCache) del(tblID, id int64) {
+	v, ok := c.elements[tblID]
+	if !ok {
+		return
+	}
+	ele, ok := v[id]
+	if !ok {
+		return
+	}
+	delete(c.elements[tblID], id)
+	c.cache.Remove(ele)
+}
+
+func (c *innerItemLruCache) put(tblID, id int64, item statistics.TableCacheItem, itemMem statistics.CacheItemMemoryUsage,
+	needEvict, needMove bool) {
+	defer func() {
+		if needEvict {
+			c.evictIfNeeded()
+		}
+	}()
+	v, ok := c.elements[tblID]
+	if !ok {
+		c.elements[tblID] = make(map[int64]*list.Element)
+		v = c.elements[tblID]
+	}
+	element, exist := v[id]
+	if exist {
+		oldItem := element.Value.(*lruCacheItem)
+		oldMemUsage := oldItem.innerMemUsage
+		oldItem.innerItem = item
+		oldItem.innerMemUsage = itemMem
+		c.calculateCost(itemMem, oldMemUsage)
+		if needMove {
+			c.cache.MoveToFront(element)
+		}
+		return
+	}
+	newItem := &lruCacheItem{
+		tblID:         tblID,
+		id:            id,
+		innerItem:     item,
+		innerMemUsage: itemMem,
+	}
+	newElement := c.cache.PushFront(newItem)
+	v[id] = newElement
+	c.calculateCost(itemMem, &statistics.IndexMemUsage{})
+}
+
+func (c *innerItemLruCache) evictIfNeeded() {
+	curr := c.cache.Back()
+	for c.trackingCost > c.capacity {
+		prev := curr.Prev()
+		item := curr.Value.(*lruCacheItem)
+		oldMem := item.innerMemUsage
+		// evict cmSketches
+		item.innerItem.DropEvicted()
+		newMem := item.innerItem.MemoryUsage()
+		c.calculateCost(newMem, oldMem)
+		// remove from lru
+		c.cache.Remove(curr)
+		delete(c.elements[item.tblID], item.id)
+		if c.onEvict != nil {
+			c.onEvict(item.tblID)
+		}
+		curr = prev
+	}
+}
+
+func (c *innerItemLruCache) calculateCost(newUsage, oldUsage statistics.CacheItemMemoryUsage) {
+	c.trackingCost += newUsage.TrackingMemUsage() - oldUsage.TrackingMemUsage()
+}
+
+func (c *innerItemLruCache) copy() *innerItemLruCache {
+	newLRU := newInnerLruCache(c.capacity)
+	curr := c.cache.Back()
+	for curr != nil {
+		item := curr.Value.(*lruCacheItem)
+		newLRU.put(item.tblID, item.id, item.innerItem, item.innerMemUsage, false, false)
+		curr = curr.Prev()
+	}
```

Comment:
Do we need to copy `c.onEvict` here?

Reply:
no,  `onEvict` will be set in `newStatsLruCache`.

Reply:
updated.